docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,221 @@
+
+The machine health checker is an integration point between node problem detection tooling expresed as node conditions and remediation to achieve a node auto repairing feature.
+
+
+### Unhealthy criteria:
+A machine/node target is unhealthy when:
+
+- The node meets the unhealthy node conditions criteria defined.
+- The Machine has no nodeRef.
+- The Machine has nodeRef but node is not found.
+- The Machine is in phase "Failed".
+
+If any of those criterias are met for longer than given timeouts, remediation is triggered.
+For the node conditions the time outs are defined by the admin. For the other cases opinionated values can be assumed.
+For a machine with no nodeRef an opinionated value could be assumed e.g 10 min.
+For a node notFound or a failed machine, the machine is considerable unrecoverable, remediation can be triggered right away.
+
+### Remediation:
+- The machine is requested for deletion.
+- The controller owning that machine, e.g machineSet reconciles towards the expected number of replicas and start the process to bring up a new machine/node tuple.
+- The machine controller drains the node.
+- The machine controller provider implementation deletes the cloud instance.
+- The machine controller deletes the machine resource.
+
+### User Stories
+
+#### Story 1
+
+- As a user I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated transparently in the case of failures.
+
+
+#### Story 2
+
+- As a cluster admin I want to define unhealthy criterias as a list of node conditions for targeted pools of machines running particular workloads.
+
+### Implementation Details/Notes/Constraints
+
+#### MachineHealthCheck CRD:
+- Enable watching a pool of machines (based on a label selector).
+- Enable defining an unhealthy node criteria (based on a list of node conditions).
+- Enable setting a threshold of unhealthy nodes. If the current number is at or above this threshold no further remediation will take place. This can be expressed as an int or as a percentage of the total targets in the pool.
+
+
+E.g:
+- I want my worker machines to be remediated when the backed node has `ready=false` or `ready=Unknown` condition for more than 10m.
+- I want remediation to temporary short-circuit if the 40% or more of the targets of this pool are unhealthy at the same time.
+
+```yaml
+apiVersion: healthchecking.cluster-api.io/v1alpha3
+kind: MachineHealthCheck
+metadata:
+  name: example
+  namespace: machine-api
+spec:
+  selector:
+    matchLabels:
+      role: worker
+  unhealthyConditions:
+  - type:    "Ready"
+    status:  "Unknown"
+    timeout: "300s"
```

Comment:
Would the implementation be expected to parse the time? If so, would it make sense to use `5m` here instead?

Reply:
This value doesn't seem to align with the description above. (5 minutes vs 10 minutes)